library-best-practices/glossary.md

```diff
@@ -0,0 +1,68 @@
+Java Dependency Glossary
+------------------------
+
+### Types of conflicts and compatibility
+
+- **Linkage conflict**: The signature, return type, modifiers, or throws
+  declaration of a non-private method or class in a dependency has changed in an
+  incompatible way between the version supplied at compile time and the version
+  invoked at runtime. For example, a public method may be removed from a class
+  or a class may be made final. Linkage conflicts detected at runtime manifest
+  as `ReflectiveOperationException`, `NoClassDefFoundError`,
+  `NoSuchFieldException`, `MethodNotFoundException`, `LinkageError`, or other
+  related exceptions. 
+  - Or, another perspective: In cases where binary compatibility and source
+    compatibility are the same, a linkage conflict is when compilation would
+    fail if the libraries in the classpath were all built together from their
+    originating source code, or when reflection would fail.
+  - Opposite: **Linkage-compatible**.
+  - Sub-type: **Static linkage conflict**: A linkage conflict caused by a direct
+    code reference (non-reflective).
+  - Sub-type: **Reflective linkage conflict**: A linkage conflict caused by a
+    reflective reference.
+
+- **Behavior conflict**: The class's implementation has changed in a way that
+  can break clients at runtime although all signatures remain compatible. For
+  example, if a method that has been returning mutable lists begins returning
+  immutable lists without updating its signature, dependents that mutate the
+  list will fail, possibly far away from the original call site. By contrast, a
+  change in return type from List to ImmutableList would be a linkage conflict.
```

Comment:
Is this a linkage conflict? Generally a return type can be made more specific without besaking existing code. Maybe flip so "change in return type from ImmutableList to List"

Reply:
Yes you're right, I will flip that.

Reply:
Wait, I didn't read carefully enough. I realized that this example is odd; it actually can break both directions. Part of the issue is that `ImmutableList` is not really substitutable for `List` even though it is derived from it.

- `List` -> `ImmutableList`: Behavior conflict, because code that used to modify the list will now throw exceptions.
- `ImmutableList` -> `List`: Static linkage conflict, because references to methods present on ImmutableList but absent from List will not be found.

We need to think of a clearer example that doesn't involve behavior conflicts.

Reply:
I don't know. Both cases are covered (with the fix) and thus this example seems OK to me. But if you don't like it, maybe an example that doesn't change the signature at all? For instance a method that used to closes an input stream it was passed, no longer does that. Or a method that used to return non-negative numbers can now return negative numbers, or something like that?

Reply:
Oy, I also didn't read the larger context of this. You're right that it is actually a good example, because it demonstrates both things. I will keep it, and fix the linkage conflict part.